prow/resultstore/payload.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resultstore
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"google.golang.org/genproto/googleapis/devtools/resultstore/v2"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	v1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+)
+
+type Payload struct {
+	Job       *v1.ProwJob
+	Started   *metadata.Started
+	Finished  *metadata.Finished
+	Files     []*resultstore.File
+	ProjectID string
+	// TODO: It is unclear whether InvocationID is needed when uploading to
+	// resultstore using the batch API. If not, invID and its functionality
+	// may be removed.
+	invID string
+}
+
+func (p *Payload) invocation() (*resultstore.Invocation, error) {
+	if p.Job == nil {
+		return nil, errors.New("internal error: Payload.Job is nil")
+	}
+	i := &resultstore.Invocation{
+		Id:                   p.invocationID(),
```

Comment:
You don't need to set `id` or `name` fields on the resource protos when using batch upload API, you only need to set UploadRequest.id.

This is based on AIP spec for create request messages, but modified to suit the needs of our system.

https://google.aip.dev/133#user-specified-ids

"The {resource}_id field must exist on the request message, not the resource itself."
"The name field on the resource must be ignored."

This goes for all the resource protos, so you can delete a bunch of your code here.

(We could probably clarify the documentation around this, but it's hard to keep clarity across all these protos.)

Reply:
In order to keep the client as simple as possible, I separated the batch upload concerns into the writer package (writer.go). writer.New() takes the InvocationId from the Invocation to create the UploadRequest and takes care of all the token exchange. There should be no trace of the batch operation in this file.

Thanks for noting that the InvocationID is not needed in any artifacts; I've removed Payload.invID now.

(Sorry, the earlier comment I wrote here was lost on github somewhere when I pushed my last commit, I think...)